core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -215,6 +220,14 @@ private[log] class ProducerAppendInfo(val producerId: Long,
 }
 
 object ProducerStateManager {
+  // If the producer's epoch wraps around, we need to be able to detect that the wrapped around epoch is
+  // "larger" than the previous one. Unfortunately, there is no guarantee that the producer will write to
+  // the same topic partitions after wrapping around, so simply checking for an epoch which has reset to
+  // 0 will not suffice. Instead we look for a wrap-around difference which is less than a fixed delta. So
+  // if the epoch has reached Short.MaxValue - n and the new epoch is m, then we will treat the new epoch
+  // as larger provided that n + m is less than this delta.
```

Comment:
It would be nice to have a description of how wrap around is meant to work. This already includes some of that information, but it starts from the edge case and explains how we handle it (which makes sense within the context of the docs). It would be a little nicer to state how it works and then talk about the edge case.